### PR TITLE
MAINT: stats.mode: revert gh-15423

### LINKED ISSUE #16418: BUG: breaking change: scipy.stats.mode returned value has changed shape without deprecation
### Describe your issue.

The scikit-learn nightly CI started to fail because of a recent change in scipy `main`:

- https://github.com/scikit-learn/scikit-learn/pull/23633#issuecomment-1157406443

In scipy `main`:

```
import numpy as np
from scipy import stats

stats.mode(np.ones(3))[0]
```

returns `1.0` while previously it would return:

```
array([1.])
```

Maybe a deprecation cycle is needed for this change?

### Reproducing Code Example

```python
import numpy as np
import scipy

scipy.stats.mode(np.ones(3))[0][0]
```

### Error message

```shell
IndexError: invalid index to scalar variable.
```

### SciPy/NumPy/Python version information

1.10.0.dev0+0.df3fe4e 1.24.0.dev0+102.gb84a53df9 sys.version_info(major=3, minor=10, micro=4, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-16418

#### What does this implement/fix?
gh-15423 changed `mode` to consume the axis it works along (like other stats reduction operations) without warning. This caused headaches downstream. This PR currently reverts gh-15423 and adds a `FutureWarning` about the default behavior change scheduled in 1.11.0.

#### Additional information
Alternative: do not revert gh-15423. Instead, complete the job: in gh-15423 we neglected to correct the axis consumption behavior for empty arrays and arrays with NaNs with `nan_policy='omit'`.

### CODE REVIEW

**Comment by tupui** [comment]
I still prefer to not do anything. Still a bug to me and it's not because we had one report saying it's concerning them that we should revert (as it has very minor impact). We had a mailing list period and some votes from our maintainers on this.

**Comment by mdhaber** [comment]
Yes, it was a bug. Ok. I'll close this unless somethunf changes. But do you agree we should then correct the axis consumption behavior for empty arrays and arrays with NaNs with nan_policy='omit'? (Before 1.9.0 releases)? Right now those are edge cases that weren't adjusted.

**Comment by mdhaber** [comment]
@ogrisel @ev-br I revived this in case you're interested.

**Comment by ogrisel** [comment]
Thanks. This looks good to me.

File: scipy/stats/_stats_py.py
Comment (h-vetinari) [inline scipy/stats/_stats_py.py:584]: The use here of single vs double backticks is inconsistent. I always forget which one too, but a mix is probably not working. 😅 

Also, there's another mention of "Support for non-numeric arrays has been deprecated and will be removed in the second release after SciPy 1.9." in the Notes that should be replaced.

Comment (mdhaber) [reply scipy/stats/_stats_py.py:584]: I'll improve the other "second release after 1.9.0" bit.

Re:backticks:, unfortunately, it is (arguably) correct usage in each case. I'm pretty consistent about applying the rules as I've interpreted them, although that is debatable.

The only guidance in the [standards](https://numpydoc.readthedocs.io/en/latest/format.html) is "Enclose variables in single backticks" and "For inline display use double backticks". 

So I've generalized these rules to:
* Functions, classes, parameters, objects, attributes, methods, and returns that are given names in our existing documentation get single backticks.
* All other text that is intended to render as "code" gets double backticks.

I don't apply these rules consistently to Python literals because I don't think everyone else agrees that, e.g., values that can be passed into functions like `True` an `1` should render as code (but I do). I would agree that eight characters is a bit excessive to type the word `True`.

Here's how I applied these rules:
`skew`, `kurtosis`, and `mode` are functions. (Note that in this case, single backticks will make links to the appropriate function documentation.)
`keepdims=True` is none of the special cases that get single backticks. Double backticks makes it render as code.
`keepdims` is the one that is questionable. It is not a parameter of the function currently, so I chose double backticks. You could argue that since it is a parameter of other functions and is planned to be a parameter of this function in the future, it should be single backticks.

Comment (h-vetinari) [reply scipy/stats/_stats_py.py:584]: I stand corrected, and thanks a lot for the very detailed answer! 🙏
Whenever I absorb this information, I tend to forget it again right away, because my brain refuses to save the significance of single-vs-double quotes of any kind (and where I do remember it took a long time). sorry 😅

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:612]:
```suggestion
        and will be removed in 1.11.0. `pandas.DataFrame.mode`_ can
```

**Review by rgommers — CHANGES_REQUESTED** [review]
Thanks @mdhaber. This doesn't look right unfortunately. A principle with both deprecation and future warnings is: there must be a way for users to fix their code. Here you are unconditionally emitting a `FutureWarning` at the top of the function, and only announce that the solution (`keepdims`) will be available in a future release.

If the solution is to add `keepdims` to `mode` (and other reductions?) then it should be done here. Absent that, I would do a straight revert for 1.9.x

**Comment by mdhaber** [comment]
That was the original intent of the PR and it's in the history. It was removed for sensible reasons. I will address those concerns when I bring it back.

**Comment by rgommers** [comment]
@mdhaber I'm not sure I understand your reply. If you don't want to add the `keepdims` that's fine, but that doesn't address my review comment. 

> Absent that, I would do a straight revert for 1.9.x

I'll elaborate if it wasn't clear. We can do this:
- Backport only the first commit in this PR to `1.9.x`
- Update the release notes, and ensure scikit-learn et al. fix their usage of `stats.mode`
- Close this PR, don't merge it.

**Comment by mdhaber** [comment]
Oops, I was replying to this part:

> If the solution is to add keepdims to mode (and other reductions?) then it should be done here

I meant the I added `keepdims` to `mode` earlier in this PR and later removed it, but I can add it back again. I'll try to do it later today.

**Comment by rgommers** [comment]
Thanks, it makes sense now. Sounds good.

**Comment by rgommers** [comment]
There were a lot of `stats` failures in CI that should all be unrelated, but merging with `main` to rerun and see if they disappear.

File: scipy/stats/_mstats_basic.py
Comment (rgommers) [inline scipy/stats/_mstats_basic.py:349]: This fix is still incomplete. With the example from the docstring of this function:
```
>>> mstats.mode(m_arr, keepdims=False)
ModeResult(mode=array([1.]), count=array([2.]))
```

**Review by rgommers — COMMENTED** [review]
CI is all green now. I pushed some minor changes to the warning and docstrings. There's still an issue with `mstats.mode` not doing the right thing for `keepdims=False`.

**Comment by mdhaber** [comment]
I think those were mostly changes in NumPy getting addressed in NumPy/numpy#21177, NumPy/numpy#22006, and scipy/scipy#16610.

File: scipy/stats/_mstats_basic.py
Comment (mdhaber) [reply scipy/stats/_mstats_basic.py:349]: It looks like you added that example to the docstring in your last commit, which is why I hadn't tested it. `keepdims` was not added as a public parameter of `stats.mstats.mode`, only `stats.mode`. That would be a separate enhancement. I

I'll remove those new examples from the docstring and raise an error if any `kwargs` other than `_keepdims` is passed into `mstats.mode`. I would prefer not to add `keepdims` as a parameter of `stats.mstats.mode` as that would require other changes and tests.

File: scipy/stats/tests/test_mstats_basic.py
Comment (mdhaber) [inline scipy/stats/tests/test_mstats_basic.py:779]: Or if preferred:
```suggestion
        message = r"mode\(\) got an unexpected keyword argument 'keepdims'"
```

File: scipy/stats/_mstats_basic.py
Comment (rgommers) [reply scipy/stats/_mstats_basic.py:351]: The way it was, the example code for `mstats.mode` is raising warnings though - that can't be right

Comment (mdhaber) [reply scipy/stats/_mstats_basic.py:351]: Yeah, I didn't notice that they were calling `stats.mode` in the docstring of `stats.mstats.mode`. No warnings are emitted by the doctests now.

File: scipy/stats/tests/test_mstats_basic.py
Comment (rgommers) [reply scipy/stats/tests/test_mstats_basic.py:779]: this seems fine as is

File: scipy/stats/_mstats_basic.py
Comment (rgommers) [reply scipy/stats/_mstats_basic.py:351]: Ah okay, I missed the underscore in `kwargs.pop("_keepdims", True)`, I thought you intended for `keepdims` to be valid. I now see the `**kwargs` is only needed because `mstats.mode` is used from within `stats.mode`.

So this should be fine as is now. Let's leave the comment unresolved in case someone gets puzzled later and revisits this PR.

**Review by rgommers — APPROVED** [review]
CI is all green, this seems to solve the problem for scikit learn (which is still at 1.1.1 and would break with the current state for scipy 1.9.0), and changes look fine/maintainable. So in it goes. Thanks a lot @mdhaber for patiently working through this.

File: scipy/stats/_mstats_basic.py
Comment (mdhaber) [reply scipy/stats/_mstats_basic.py:351]: Another way I could have done it is to move the implementation of `mstats.mode` into a private method called by both `mstats.mode` and `stats.mode`. that would keep `kwargs` out of the signature and avoid the need to manually produce the unexpected argument error. Want me to refactor?

Comment (rgommers) [reply scipy/stats/_mstats_basic.py:351]: Sure, whatever you prefer. Let's leave that out of `1.9.x` indeed.

**Comment by glemaitre** [comment]
Thanks to all.

**Pull request merged**
